// gitnexus/dist/storage/repo-manager.js
/**
 * Repository Manager
 *
 * Manages GitNexus index storage in .gitnexus/ at repo root.
 * Also maintains a global registry at ~/.gitnexus/registry.json
 * so the MCP server can discover indexed repos from any cwd.
 */
import fs from 'fs/promises';
import path from 'path';
import os from 'os';
const GITNEXUS_DIR = '.gitnexus';
// ─── Local Storage Helpers ─────────────────────────────────────────────
/**
 * Get the .gitnexus storage path for a repository
 */
export const getStoragePath = (repoPath) => {
    return path.join(path.resolve(repoPath), GITNEXUS_DIR);
};
/**
 * Get paths to key storage files
 */
export const getStoragePaths = (repoPath) => {
    const storagePath = getStoragePath(repoPath);
    return {
        storagePath,
        kuzuPath: path.join(storagePath, 'kuzu'),
        metaPath: path.join(storagePath, 'meta.json'),
    };
};
/**
 * Load metadata from an indexed repo
 */
export const loadMeta = async (storagePath) => {
    try {
        const metaPath = path.join(storagePath, 'meta.json');
        const raw = await fs.readFile(metaPath, 'utf-8');
        return JSON.parse(raw);
    }
    catch {
        return null;
    }
};
/**
 * Save metadata to storage
 */
export const saveMeta = async (storagePath, meta) => {
    await fs.mkdir(storagePath, { recursive: true });
    const metaPath = path.join(storagePath, 'meta.json');
    await fs.writeFile(metaPath, JSON.stringify(meta, null, 2), 'utf-8');
};
/**
 * Check if a path has a GitNexus index
 */
export const hasIndex = async (repoPath) => {
    const { metaPath } = getStoragePaths(repoPath);
    try {
        await fs.access(metaPath);
        return true;
    }
    catch {
        return false;
    }
};
/**
 * Load an indexed repo from a path
 */
export const loadRepo = async (repoPath) => {
    const paths = getStoragePaths(repoPath);
    const meta = await loadMeta(paths.storagePath);
    if (!meta)
        return null;
    return {
        repoPath: path.resolve(repoPath),
        ...paths,
        meta,
    };
};
/**
 * Find .gitnexus by walking up from a starting path
 */
export const findRepo = async (startPath) => {
    let current = path.resolve(startPath);
    const root = path.parse(current).root;
    while (current !== root) {
        const repo = await loadRepo(current);
        if (repo)
            return repo;
        current = path.dirname(current);
    }
    return null;
};
/**
 * Add .gitnexus to .gitignore if not already present
 */
export const addToGitignore = async (repoPath) => {
    const gitignorePath = path.join(repoPath, '.gitignore');
    try {
        const content = await fs.readFile(gitignorePath, 'utf-8');
        if (content.includes(GITNEXUS_DIR))
            return;
        const newContent = content.endsWith('\n')
            ? `${content}${GITNEXUS_DIR}\n`
            : `${content}\n${GITNEXUS_DIR}\n`;
        await fs.writeFile(gitignorePath, newContent, 'utf-8');
    }
    catch {
        // .gitignore doesn't exist, create it
        await fs.writeFile(gitignorePath, `${GITNEXUS_DIR}\n`, 'utf-8');
    }
};
// ─── Global Registry (~/.gitnexus/registry.json) ───────────────────────
/**
 * Get the path to the global GitNexus directory
 */
export const getGlobalDir = () => {
    return path.join(os.homedir(), '.gitnexus');
};
/**
 * Get the path to the global registry file
 */
export const getGlobalRegistryPath = () => {
    return path.join(getGlobalDir(), 'registry.json');
};
/**
 * Read the global registry. Returns empty array if not found.
 */
export const readRegistry = async () => {
    try {
        const raw = await fs.readFile(getGlobalRegistryPath(), 'utf-8');
        const data = JSON.parse(raw);
        return Array.isArray(data) ? data : [];
    }
    catch {
        return [];
    }
};
/**
 * Write the global registry to disk
 */
const writeRegistry = async (entries) => {
    const dir = getGlobalDir();
    await fs.mkdir(dir, { recursive: true });
    await fs.writeFile(getGlobalRegistryPath(), JSON.stringify(entries, null, 2), 'utf-8');
};
/**
 * Register (add or update) a repo in the global registry.
 * Called after `gitnexus analyze` completes.
 */
export const registerRepo = async (repoPath, meta) => {
    const resolved = path.resolve(repoPath);
    const name = path.basename(resolved);
    const { storagePath } = getStoragePaths(resolved);
    const entries = await readRegistry();
    const existing = entries.findIndex((e) => path.resolve(e.path) === resolved);
    const entry = {
        name,
        path: resolved,
        storagePath,
        indexedAt: meta.indexedAt,
        lastCommit: meta.lastCommit,
        stats: meta.stats,
    };
    if (existing >= 0) {
        entries[existing] = entry;
    }
    else {
        entries.push(entry);
    }
    await writeRegistry(entries);
};
/**
 * Remove a repo from the global registry.
 * Called after `gitnexus clean`.
 */
export const unregisterRepo = async (repoPath) => {
    const resolved = path.resolve(repoPath);
    const entries = await readRegistry();
    const filtered = entries.filter((e) => path.resolve(e.path) !== resolved);
    await writeRegistry(filtered);
};
/**
 * List all registered repos from the global registry.
 * Optionally validates that each entry's .gitnexus/ still exists.
 */
export const listRegisteredRepos = async (opts) => {
    const entries = await readRegistry();
    if (!opts?.validate)
        return entries;
    // Validate each entry still has a .gitnexus/ directory
    const valid = [];
    for (const entry of entries) {
        try {
            await fs.access(path.join(entry.storagePath, 'meta.json'));
            valid.push(entry);
        }
        catch {
            // Index no longer exists — skip
        }
    }
    // If we pruned any entries, save the cleaned registry
    if (valid.length !== entries.length) {
        await writeRegistry(valid);
    }
    return valid;
};
/**
 * Get the path to the global CLI config file
 */
export const getGlobalConfigPath = () => {
    return path.join(getGlobalDir(), 'config.json');
};
/**
 * Load CLI config from ~/.gitnexus/config.json
 */
export const loadCLIConfig = async () => {
    try {
        const raw = await fs.readFile(getGlobalConfigPath(), 'utf-8');
        return JSON.parse(raw);
    }
    catch {
        return {};
    }
};
/**
 * Save CLI config to ~/.gitnexus/config.json
 */
export const saveCLIConfig = async (config) => {
    const dir = getGlobalDir();
    await fs.mkdir(dir, { recursive: true });
    await fs.writeFile(getGlobalConfigPath(), JSON.stringify(config, null, 2), 'utf-8');
};
